be/src/exprs/agg/window.h

```diff
@@ -440,24 +464,53 @@ class LeadLagWindowFunction final : public ValueWindowFunction<LT, LeadLagState<
             return;
         }
 
-        if (!columns[0]->is_null(frame_end - 1)) {
-            this->data(state).is_null = false;
-            const Column* data_column = ColumnHelper::get_data_column(columns[0]);
-            const InputColumnType* column = down_cast<const InputColumnType*>(data_column);
-            AggDataTypeTraits<LT>::assign_value(this->data(state).value,
-                                                AggDataTypeTraits<LT>::get_row_ref(*column, frame_end - 1));
-        } else {
-            if (!ignoreNulls) {
-                this->data(state).is_null = true;
-                return;
+        // for lead/lag, [peer_group_start, peer_group_end] equals to [partition_start, partition_end]
+        // when lead/lag called, the whole partitoin's data has already been here, so we can just check all the way to the begining or the end
+        if (ignoreNulls) {
+            // lead(v1 ignore nulls, <offset>) has window `ROWS BETWEEN UNBOUNDED PRECEDING AND <offset> FOLLOWING`
+            //      frame_start = partition_start
+            //      frame_end = current_row + <offset> + 1
```

Comment:
Why `frame_end ` depends `current_row ` ? `frame_end ` is partition/window's end?

Reply:
frame_end means window'end, and peer_group_end means partition_end